File: app/src/main/java/com/example/fruitsdiary/data/fruit/FruitDataSource.java
package com.example.fruitsdiary.data.fruit;

import com.example.fruitsdiary.data.DataSource;
import com.example.fruitsdiary.model.Fruit;
import com.example.fruitsdiary.network.FruitsDiaryService;
import com.example.fruitsdiary.util.SchedulerProvider;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observable;


public class FruitDataSource extends DataSource {

    @Inject
    FruitDataSource(FruitsDiaryService service, SchedulerProvider provider) {
        super(service, provider);
    }

    public Observable<List<Fruit>> getFruits() {
        return mService.getAllFruits()
                .subscribeOn(mProvider.io())
                .observeOn(mProvider.ui());
    }
}

File: app/src/main/java/com/example/fruitsdiary/data/fruit/FruitDataSource.kt
package com.example.fruitsdiary.data.fruit

import com.example.fruitsdiary.data.DataSource
import com.example.fruitsdiary.model.Fruit
import com.example.fruitsdiary.network.FruitsDiaryService
import com.example.fruitsdiary.util.SchedulerProvider
import io.reactivex.Observable
import javax.inject.Inject


class FruitDataSource @Inject
internal constructor(service: FruitsDiaryService, provider: SchedulerProvider) : DataSource(service, provider) {

    fun getFruits(): Observable<List<Fruit>> =
            service.getAllFruits()
                    .subscribeOn(provider.io())
                    .observeOn(provider.ui())
}
